src/openms/source/FILTERING/DATAREDUCTION/IsotopeDistributionCache.cpp

```diff
@@ -40,89 +40,116 @@
+    //determine the number of optional peaks at the beginning/end
+    Size begin = 0;
+    Size end = 0;
+    bool is_begin = true;
+    bool is_end = false;
+    double max = 0.0;
+    for (Size i = 0; i < isotopes.intensity.size(); ++i)
+    {
+      if (isotopes.intensity[i] < intensity_percentage_)
       {
-        isotope_distributions_[index].intensity.push_back(it->getIntensity());
-        //log_ << " - " << it->second << std::endl;
+        if (!is_end && !is_begin) is_end = true;
+        if (is_begin) ++begin;
+        else if (is_end) ++end;
       }
-
-      //determine the number of optional peaks at the beginning/end
-      Size begin = 0;
-      Size end = 0;
-      bool is_begin = true;
-      bool is_end = false;
-      for (Size i = 0; i < isotope_distributions_[index].intensity.size(); ++i)
+      else if (is_begin)
       {
-        if (isotope_distributions_[index].intensity[i] < intensity_percentage)
-        {
-          if (!is_end && !is_begin)
-            is_end = true;
-          if (is_begin)
-            ++begin;
-          else if (is_end)
-            ++end;
-        }
-        else if (is_begin)
-        {
-          is_begin = false;
-        }
+        is_begin = false;
       }
-      isotope_distributions_[index].optional_begin = begin;
-      isotope_distributions_[index].optional_end = end;
 
       //scale the distribution to a maximum of 1
-      double max = 0.0;
-      for (Size i = 0; i < isotope_distributions_[index].intensity.size(); ++i)
+      if (isotopes.intensity[i] > max)
       {
-        if (isotope_distributions_[index].intensity[i] > max)
-        {
-          max = isotope_distributions_[index].intensity[i];
-        }
+        max = isotopes.intensity[i];
       }
+    }
+    isotopes.optional_begin = begin;
+    isotopes.optional_end = end;
 
-      isotope_distributions_[index].max = max;
+    isotopes.max = max;
 
-      for (Size i = 0; i < isotope_distributions_[index].intensity.size(); ++i)
-      {
-        isotope_distributions_[index].intensity[i] /= max;
-      }
+    for (Size i = 0; i < isotopes.intensity.size(); ++i)
+    {
+      isotopes.intensity[i] /= max;
     }
   }
 
   // Returns the isotope distribution for a certain mass window
-  const IsotopeDistributionCache::TheoreticalIsotopePattern& IsotopeDistributionCache::getIsotopeDistribution(double mass) const
+  const IsotopeDistributionCache::TheoreticalIsotopePattern& IsotopeDistributionCache::getIsotopeDistribution(double mass)
   {
-    //calculate index in the vector
-    Size index = static_cast<Size>(std::floor(mass / mass_window_width_));
-
+    Size index = Size (std::floor(mass/mass_window_width_));
     if (index >= isotope_distributions_.size())
     {
-      throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "IsotopeDistribution not precalculated. Maximum allowed index is " + String(isotope_distributions_.size()), String(index));
+      Size size = isotope_distributions_.size();
```

Comment:
this code should move into `calculateisotopeDistribution`.
simply pass the new maximum index which you need to have filled.
The rest should happen in the function.
Avoids code duplication (see https://github.com/cbielow/OpenMS/pull/92/commits/b2b8728926d23591a158b56ed31ee4b9f5e16c59#diff-8d025ab465e7de035c1eae4c0d2440b9R145)